### DISCUSSION TITLE
how to add word to the start of select multi line

### DISCUSSION THREAD

**Discussion by 668168** [Taitava/obsidian-shellcommands#207] [Q&A] [answered]
for example, to add a list with `-`:

Before:

shellcommand is amazing.
wonderful  tools for us.
thanks very mach.

After:
-. shellcommand is amazing.
-. wonderful  tools for us.
-. thanks very mach.

**Comment by Taitava** [comment, marked as answer]
Hmm, this is an interesting question. 🙂 

To be able to better answer this, I'd need to know which operating system you are using. For now, I'll just assume you are using macOS or Linux - but if you are using Windows, let me know and I'll come up with another solution that should work for Windows.

(I moved this discussion from _Ideas_ to _Q & A_).

# macOS/Linux solution

1. Define the following shell command: `printf {{selection}} | sed -zE 's/(^|\n)/\1 - /g'`
2. Define an _alias_ for the shell command, e.g. _Make list_
3. Set the `stdout` output channel to be _Current file: caret position_.
4. Select some text in a note, open the command palette, and run the command _Execute: Make list_.

Btw. I assume that you meant to use `-` instead of `-.` but if you need the dot too, just add it after the dash `-` in the shell command.

The shell command is quite complex to explain throughout, but I'll try to cover it roughly:
1. `printf {{selection}} | ` is used to pipe the selected text to the next command, which appears after the pipe character `|`.
2. `sed -zE 's/(^|\n)/\1 - /g'` does a search & replace operation using a [regular expression](https://en.wikipedia.org/wiki/Regular_expression), which can be a bit hard topic to learn if you don't know it yet.
    - `(^|\n)` is what we are searching for: we search for the beginning of the text, and also for each newline characters (denoted with `\n`).
    - `\1 - ` is the replacement. For the first line, we will just add ` - ` at the beginning, and for all other lines, we are replacing a newline character with a newline **plus** ` - `.
3. With the output channel set to _Current file: caret position_, we define that the output should replace the selected text in Obsidian.

The explanation was really complex because I wanted to make it short. Please let me know if you have any questions on some details, and I'll explain more.

**Reply by 668168** [reply]
wow, so nice ,so detail explain ~~, best regard . 
yes , i use macbook. friendly for *nux develop and bash shell . 
before i post the question, i first idea is sed with Reqular expression too. 
but i am pain with sed edit for much years /cry and lazy to study sed...., and I think, may be there is some other best saturation . such as  sublime,vim with RE. or use python with RE, etc.

follow your suggest , may be I need to take some time to learn sed command . /cry . because may be  sed is the only edit tools in shell environment .

**Reply by Taitava** [reply]
To be honest, I haven't used `sed` much myself. I'm familiar with regular expressions (from PHP, JavaScript and little bit of Python), so it helps me to understand the `sed` usage for the most parts, but I did need to do some research before I was able to provide the answer.

This was a really nice question to answer to, because I hadn't thought previously about using shell commands to add some text in front of all selected lines! 🙂 It was nice to find out that it actually worked quite well! 🥂

**Reply by 668168** [reply]
yes, just for funny. 
it is a SC tools, i has use it for many basic use and  want to try some advance  useage. it is usefull and wonderful .
